test: add EngineApiTreeHandlerImpl unit test

fgimenez opened the pull request:
Follow up of #9365 I've added a `PersistenceHandler` trait to be able to insert a Mock in the tree and do some assertions after persisting blocks.

Rjected reviewed (changes requested):
I'm not sure we should introduce a trait for the sender just for testing, I would prefer a test task for mocking responses instead, wdyt?

fgimenez commented:
> I'm not sure we should introduce a trait for the sender just for testing, 

sorry for not being clear in the PR description. as I see it it is the other way around, the trait allows us to decouple components, and this decoupling enables unit testing. otherwise `EngineApiTreeHandlerImpl` would be tightly coupled to `PersistenceHandler` and could not be unit tested without testing `PersistenceHandler` too.

in my view the trait introduced in this PR is similar to the current `NetworkSyncUpdater` trait used in the current `BeaconConsensusEngine` here https://github.com/paradigmxyz/reth/blob/main/crates/consensus/beacon/src/engine/mod.rs#L187 It is only implemented by `NetworkHandle` here https://github.com/paradigmxyz/reth/blob/main/crates/net/network/src/network.rs#L368, used in prod code, and by `NoopSyncStateUpdater` here https://github.com/paradigmxyz/reth/blob/main/crates/net/p2p/src/sync.rs#L57, which is only used in tests.

anyway, happy to discard these changes and test the functionality at other levels, lmk wdyt

> I would prefer a test task for mocking responses instead, wdyt?

that would be more like an integration test right? I was planning to add it too in a separate PR.

as I see it it is

mattsse reviewed (changes requested):
love the tests, but I think we can solve testing this without making the handle a trait

fgimenez commented:
ok, i've removed the trait and updated the test, it doesn't use a separate `PersistenceHandlerMock` type, just a `PersistenceAction` channel that allows instantiating the `PersistenceHandler` and assert on the received actions. @mattsse @Rjected ptal

mattsse commented on crates/engine/tree/src/test_utils.rs:
```diff
@@ -75,3 +81,33 @@ pub(crate) fn insert_headers_into_client(
+    let sealed_with_senders = SealedBlockWithSenders::new(sealed.clone(), vec![sender]).unwrap();
+
+    ExecutedBlock::new(
+        Arc::new(sealed),
+        Arc::new(sealed_with_senders.senders),
+        Arc::new(ExecutionOutcome::new(
+            BundleState::default(),
+            Receipts { receipt_vec: vec![] },
+            block_number,
+            vec![Requests::default()],
+        )),
+        Arc::new(HashedPostState::default()),
+        Arc::new(TrieUpdates::default()),
```
making mental note here, that we should revisit what needs to be arc and what doesn't or what can be grouped

mattsse reviewed (approved):
lgtm!

Rjected reviewed (approved):
lgtm!
